Internal Solution for Why Did the Cow Cross the Road (Gold) (#1953)
cpinitiative/usaco-guide: solutions/usaco-718.mdx

---
id: usaco-718
source: USACO Gold 2017 February
title: Why Did the Cow Cross The Road II
author: Brad Ma
---
 
[Official Analysis (C++)](http://www.usaco.org/current/data/sol_nocross_gold_feb17.html)

The recursive approach where you try all configurations of crosswalks has the 
time complexity $\mathcal{O}(2 ^ N)$, which is extremely slow. We observe that we have 

> many duplicate recursive calls, so by using a top down dynamic progamming approach with 

Reviewer: would be nice to explain what is being duplicated

memoization we can significantly reduce the time complexity to $\mathcal{O}(N ^ 2)$.
In this approach, we store our answers such that if there ever is a call to it again, the

> answer just uses the cached value instead of recomputing it. A bottom up dynamic progamming 

Reviewer 2: ```suggestion
function uses the cached answer instead of recomputing it. A bottom up dynamic progamming 
```

---
id: usaco-718
source: USACO Gold 2017 February
title: Why Did the Cow Cross The Road II
author: Brad Ma
---
 
[Official Analysis (C++)](http://www.usaco.org/current/data/sol_nocross_gold_feb17.html)

> The normal recursive approach where you try all configurations of crosswalks has the 

Reviewer 3: ```suggestion
The recursive approach where you try all configurations of crosswalks has the 
```
is normal necessary here?

> time complexity $\mathcal{O}(2 ^ N)$ which is extremely slow. We observe that we have 

Reviewer 2: ```suggestion
time complexity $\mathcal{O}(2 ^ N)$, which is extremely slow. We observe that we have 
```

many duplicate recursive calls, so by using a top down dynamic progamming approach with 
memoization we can significantly reduce the time complexity to $\mathcal{O}(N ^ 2)$.

> In this approach, we just store our answers so if there ever is a call to it again, the

Reviewer: ```suggestion
In this approach, we store our answers such that if there ever is a call to it again, the
```

> the time complexity to $\mathcal{O}(N ^ 2)$. In this approach, we store our answers such 

Reviewer 4: Maybe add a paragraph break here? Otherwise, it looks good

> answer is just returned. A bottom up dynamic progamming approach without recursion can 

Reviewer: ```suggestion
answer is just uses the cached value instead of recomputing it. A bottom up dynamic progamming approach without recursion can 
```

> answer is just returned. A bottom up dynamic progamming approach without recursion can 

Reviewer: is mentioning bottom up really necessary

Author: yes because you can do dp with recursion + memorization or dp without recursion

recomputing it. A bottom up dynamic progamming approach without recursion can run faster 
because function calls have a large constant, but it is more difficult to implement and 
is not necessary to pass all test cases.

## Implementation

**Time Complexity:** $\mathcal{O}(N ^ 2)$

<LanguageSection>
  
<JavaSection>
  
```java
import java.io.*;
import java.util.*;

> public class WhyDidTheCowCrossTheRoadII {

Reviewer 5: make this one NoCross won't you?

Author: i will just so you approve, but the 5 other reviewers didn't have a problem with this

	static int[] firstSide;
	static int[] secondSide;
	static int[][] memoization;

> 	static boolean friendly (int n, int m) {

Reviewer 5: jesus heck who puts a space between the function names and the paramater list

		if (Math.abs(firstSide[n] - secondSide[m]) <= 4) {
			return true;
		}

> 		return false;

Reviewer 5: just `return Math.abs(firstSide[n] - secondSide[n]) <= 4;`

Author: that will make it less readable. it isn't always about making it the shortest lines

	}

> 	static int solve (int n, int m) {

Reviewer 5: what's n and m lol

			return 0;
		}
		// if already computed, just return the value
		if (memoization[n][m] != -1) {  
			return memoization[n][m];
		}

		if (friendly(n, m)) {  // they are friendly, connect
			int answer =  1 + solve(n - 1, m - 1);
			memoization[n][m] = answer;
			return answer;
		}

> 		// they are not friendly, you can move one or the other, take the max

Reviewer 2: I think "move one or the other" might be a bit unclear.

Reviewer 2: Maybe say something like "we can skip either the cow on the left side or the cow on the right"
